libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp

```diff
@@ -31,19 +31,28 @@
 #include <AP_SerialManager/AP_SerialManager.h>
 #include <ctype.h>
 #include <stdio.h>
+#include <GCS_MAVLink/GCS.h>
 
 #define RP_DEBUG_LEVEL 0
 
+/*
+ * DEBUG macros:
+ * Level 1 sends message to GCS.
+ * Level 2 sends message to UART5 (pixhawk console output)
+ * Level 3 sends message to both.
+ * Choose "_ts" version in "Debug" macro to append timestamp.
+ * WARNING! Sending too much data creates significant delays (especially via UART)!
+ */
 #if RP_DEBUG_LEVEL
-  #include <GCS_MAVLink/GCS.h>
-/* Send debug messages via Mavlink to Ground Station */
-  #define Debug_gcs(level, fmt, args ...)  do { if (level <= RP_DEBUG_LEVEL) { gcs().send_text(MAV_SEVERITY_INFO, "RPLIDAR: " fmt, ## args); } } while (0)
-/* Send debug messages via Mavlink to Ground Station and append timestamp */
-  #define Debug_gcs_ts(level, fmt, args ...)  do { if (level <= RP_DEBUG_LEVEL) { gcs().send_text(MAV_SEVERITY_INFO, "%dms/RPLIDAR: " fmt, AP_HAL::millis(), ## args); } } while (0)
-/* Send debug messages to UART0 port (on pixhawk console output) and append timestamp */
-  #define Debug_uart(level, fmt, args ...)  do { if (level <= RP_DEBUG_LEVEL) { printf( "%dms/RPLIDAR: " fmt "\n", AP_HAL::millis(), ## args); } } while (0)
 /* Redirect debug messages to whatever output is needed */
-  #define Debug(level, fmt, args ...)  Debug_gcs_ts(level, fmt, ## args); Debug_uart(level, fmt, ## args);
+  #define Debug(level, fmt, args ...)  Debug_gcs(level, fmt, ## args); Debug_uart(level, fmt, ## args);
+/* Send debug messages via Mavlink to Ground Station (_ts version appends timestamp) */
+  #define Debug_gcs(level, fmt, args ...)  do { if (level&0x01) { gcs().send_text(MAV_SEVERITY_INFO, "RPL: " fmt, ## args); } } while (0)
```

Comment:
What are you trying to do with the &0x01 here? why not use level <= RP_DEBUG_LEVEL ?

Reply:
That's the way to control the destination of a message - I can easily specify if the message should be sent to GCS (0x01) or UART (0x02) port (or both - 0x03). For me it was just more useful than setting different debugging levels with the same output. Especially that printing too much data takes lots of time (effectively blocking the program! Even though it should be asynchronous, shouldn't it!?) so I had to be careful with it.

Reply:
The best way is I guess to keep both options. Like you had in one of your versions: One macro for each port, all macros accepting DEBUG_LEVEL. That way is very easy to turn on/off the debug and the verbosity level in a simple parameter. The other Device drivers in ArduPilot do it like this, so it makes it easier to understand.

Reply:
Yeah, why not. Will do that tomorrow.